replicate/common/generic-replicator.go

```diff
@@ -198,6 +200,50 @@ func (r *GenericReplicator) NamespaceAdded(ns *v1.Namespace) {
 	}
 }
 
+// NamespaceUpdated checks if namespace's labels changed and deletes any 'replicate-to-matching' resources
+// the namespace no longer qualifies for. Then it attempts to replicate resources into the updated ns based
+// on the updated set of labels
+func (r *GenericReplicator) NamespaceUpdated(nsOld *v1.Namespace, nsNew *v1.Namespace) {
+	logger := log.WithField("kind", r.Kind).WithField("target", nsNew.Name)
+	// check if labels changed
+	labelsEqual := reflect.DeepEqual(nsNew.Labels, nsOld.Labels)
+	if  labelsEqual{
+		logger.Debug("labels didn't change")
+	} else {
+		// delete any resources where namespace labels no longer match
+		oldLabels := GetKeysFromStringMap(nsOld.Labels)
+		logger.Debugf("old labels of ns %s : %v", nsOld.Name, oldLabels)
+		newLabels := GetKeysFromStringMap(nsNew.Labels)
+		logger.Debugf("new Labels of ns %s : %v", nsNew.Name, newLabels)
+		removedLabelsKeys := GetDifferenceBetweenStringLists(oldLabels, newLabels)
+		logger.Debugf("removed labels of ns %s: %v", nsNew.Name, removedLabelsKeys)
+		deletedLabels := labels.Set{}
+		for _, key := range removedLabelsKeys {
+			deletedLabels[key] = nsOld.Labels[key]
+		}
+		// check 'replicate-to-matching' resources against the deleted labels
+		for sourceKey, selector := range r.ReplicateToMatchingList {
+			if selector.Matches(deletedLabels) {
```

Comment:
I'm unsure this will work when the label selector consists of multiple labels. Example:

1. Namespace has labels `a,b,c`, Secret has replication selector for `a,b,c`
2. Namespace is updated to have labels `a,b`
3. `deletedLabels` will be `c`
4. In this case, `selector` (for `a,b,c`) will _not_ match `deletedLabels` and the replicated secrets will not be deleted (although they should be).

Shouldn't this lead to objects _not_ being deleted, even if their original replication label selectors do not match anymore (i.e. only match partially anymore)?

Reply:
Yeah, I wasn't fully aware of how selectors work, I'll figure out a solution

Reply:
Should work properly now